tokio-util/src/sync/cancellation_token/maybe_dangling.rs

```diff
@@ -0,0 +1,45 @@
+use core::future::Future;
+use core::mem::MaybeUninit;
+use core::pin::Pin;
+use core::ptr;
+use core::task::{Context, Poll};
+
+/// A wrapper type that avoids implicitly asserting the contents are currently valid.
+/// It's safe to contain a dangling reference, as long as it is not used.
+///
+/// # Safety
+///
+/// The inner type must be dropped before the data it references to.
+///
```

Comment:
It doesn't really make much sense for this to have a safety comment. There are no safety requirements for the user; in fact, using this type gives the user _more_ abilities than they usually have. Safety comments always talk about reductions in what the user can do.
```suggestion
/// A wrapper type that tells the compiler that the contents might not be valid.
///
/// This is necessary mainly when `T` contains a reference. In that case, the
/// compiler will sometimes assume that the reference is always valid; in some
/// cases it will assume this even after the destructor of `T` runs. For
/// example, when a reference is used as a function argument, then the compiler
/// will assume that the reference is valid until the function returns, even if
/// the reference is destroyed during the function. When the reference is used
/// as part of a self-referential struct, that assumption can be false. Wrapping
/// the reference in this type prevents the compiler from making that
/// assumption.
///
/// # Invariants
///
/// The `MaybeUninit` will always contain a valid value until the destructor runs.
```

Reply:
Oh, totally makes sense